From dba4fb464a72f678740b07f781a5f57924a958d0 Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Thu, 16 Jul 2020 17:38:29 +0800
Subject: [PATCH] drm/rockchip: reset conn->state->best_encoder

reset conn->state->best_encoder, otherwise will dump the following
warning log at one crtc multiple connector platform:

WARNING: CPU: 6 PID: 61 at drivers/gpu/drm/drm_atomic_helper.c:225
set_best_encoder.isra.3+0x40/0xf0

[    1.520856] Call trace:
[    1.520868]  set_best_encoder.isra.3+0x40/0xf0
[    1.520896]  drm_atomic_helper_check_modeset+0x670/0xa60
[    1.520907]  drm_atomic_helper_check+0x1c/0xc0
[    1.520920]  drm_atomic_check_only+0x354/0x640
[    1.520932]  drm_atomic_commit+0x18/0x58
[    1.520960]  rockchip_drm_bind+0x12f0/0x18e8
[    1.520973]  try_to_bring_up_master+0x144/0x1a8
[    1.520985]  component_master_add_with_match+0xd4/0x130
[    1.520996]  rockchip_drm_platform_probe+0x178/0x2e8
[    1.521022]  platform_drv_probe+0x50/0xa8
[    1.521033]  really_probe+0x200/0x2b0
[    1.521043]  driver_probe_device+0x58/0x100
[    1.521054]  __device_attach_driver+0x90/0xc0
[    1.521064]  bus_for_each_drv+0x70/0xc8
[    1.521074]  __device_attach+0xdc/0x138
[    1.521097]  device_initial_probe+0x10/0x18
[    1.521107]  bus_probe_device+0x94/0xa0
[    1.521118]  deferred_probe_work_func+0x80/0xb8
[    1.521131]  process_one_work+0x1a0/0x418
[    1.521143]  worker_thread+0x1e4/0x3e0
[    1.521154]  kthread+0x12c/0x158
[    1.521166]  ret_from_fork+0x10/0x18
[    1.521189] ---[ end trace bb121b6492800165 ]---

Change-Id: I64a35a13819add55a565165198b1313f8cfc7597
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_drv.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
index 5fe1d3897717..8cd6a59afde2 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
@@ -808,7 +808,7 @@ static void show_loader_logo(struct drm_device *drm_dev)
 	struct list_head mode_set_list;
 	struct list_head mode_unset_list;
 	unsigned int plane_mask = 0;
-	int ret;
+	int ret, i;
 
 	root = of_get_child_by_name(np, "route");
 	if (!root) {
@@ -915,6 +915,12 @@ static void show_loader_logo(struct drm_device *drm_dev)
 		 */
 		WARN_ON(update_state(drm_dev, state, set, &plane_mask));
 
+	for (i = 0; i < state->num_connector; i++) {
+		if (state->connectors[i].new_state->connector->status !=
+		    connector_status_connected)
+			state->connectors[i].new_state->best_encoder = NULL;
+	}
+
 	ret = drm_atomic_commit(state);
 	/**
 	 * todo
-- 
2.35.3

